### PR TITLE
Use ExecutorService/Future to ensure Yarn KillApplication() call timeout

### PR DESCRIPTION
**Pull request by tsangz2013**
**Why we need this**
We still observe that the Yarn KillApplication() call got stuck for ~ 15 minutes before it throws exception, so we need a way to mitigate to avoid a failed job stuck for too long and thus breaking the SLA, on Azkaban system side.

**What's done**
Use Future to handle each Yarn KillApplication() call, and add a timeout to each Future usage, so as it will only wait for a short period of time.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/utils/YarnUtils.java [hunk 1]
```diff
@@ -107,17 +112,32 @@ public static List<ApplicationReport> getAllAliveAppReportsByExecIDs(final YarnC
...
-   * Uses YarnClient to kill the jobs one by one
+   * Uses YarnClient to kill the jobs one by one, each kill has a timeout
    */
   public static void killAllAppsOnCluster(YarnClient yarnClient, Set<String> applicationIDs,
       Logger log) {
     log.info(String.format("Killing applications: %s", applicationIDs));
+    ExecutorService executor = Executors.newSingleThreadExecutor();
 
     for (final String appId : applicationIDs) {
+      Future<?> future = executor.submit(new Runnable() {
```
Comment (NancyXie2022) [inline azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:123]: Has this been tested?

File: azkaban-common/src/main/java/azkaban/utils/YarnUtils.java [hunk 2]
```diff
@@ -107,17 +112,32 @@ public static List<ApplicationReport> getAllAliveAppReportsByExecIDs(final YarnC
 
 
   /**
-   * Uses YarnClient to kill the jobs one by one
+   * Uses YarnClient to kill the jobs one by one, each kill has a timeout
    */
   public static void killAllAppsOnCluster(YarnClient yarnClient, Set<String> applicationIDs,
       Logger log) {
     log.info(String.format("Killing applications: %s", applicationIDs));
+    ExecutorService executor = Executors.newSingleThreadExecutor();
```
Comment (svnarumugam) [inline azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:120]: I am curious to know the reasoning behind single thread executor. would it be helpful if we use a multiple threads?

File: azkaban-common/src/main/java/azkaban/utils/YarnUtils.java [hunk 3]
```diff
@@ -107,17 +112,32 @@ public static List<ApplicationReport> getAllAliveAppReportsByExecIDs(final YarnC
...
+            log.warn("something happened while trying to kill this job: " + appId, t);
+          }
+        }
+      });
+      // wait for the kill with a timeout
       try {
-        YarnUtils.killAppOnCluster(yarnClient, appId, log);
-      } catch (final Throwable t) {
-        log.warn("something happened while trying to kill this job: " + appId, t);
+        future.get(YARN_APP_TIMEOUT_IN_MILLIONSECONDS, TimeUnit.MILLISECONDS);
```
Comment (svnarumugam) [inline azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:135]: Can we consider adding a unit test for the killAllAppscluster method?

**Review by svnarumugam — COMMENTED** [review]
Thanks @tsangz2013 for putting this change together! Left a few comments. Changes look good.

File: azkaban-common/src/main/java/azkaban/utils/YarnUtils.java (hunk 1 above)
Comment (tsangz2013) [reply azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:123]: let me add some unit tests

File: azkaban-common/src/main/java/azkaban/utils/YarnUtils.java (hunk 3 above)
Comment (tsangz2013) [reply azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:135]: I couldn't find a way to make a unit test work: mock sleeping in unit test is hard, instead I tested in an executor pod, the killing works well and the future timed out after 5 seconds

File: azkaban-common/src/main/java/azkaban/utils/YarnUtils.java (hunk 1 above)
Comment (NancyXie2022) [reply azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:123]: Do you mind put the execution link here? and any log line saying about the timeout?

Comment (tsangz2013) [reply azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:123]: I posted in our internal channel

Comment (NancyXie2022) [reply azkaban-common/src/main/java/azkaban/utils/YarnUtils.java:123]: 13-12-2022 11:46:54 PST pigWordCount1 WARN - Timed out killing the yarn app application_1670314587930_1089113, cancelling the runnable...

**Review by NancyXie2022 — APPROVED** [review]

**Pull request merged**